receiver/awsfirehosereceiver/internal/unmarshaler/cwlog/unmarshaler.go

```diff
@@ -4,101 +4,130 @@
-	for recordIndex, compressedRecord := range records {
-		record, err := compression.Unzip(compressedRecord)
-		if err != nil {
-			u.logger.Error("Failed to unzip record",
+func (u *Unmarshaler) UnmarshalLogs(compressedRecord []byte) (plog.Logs, error) {
+	var err error
+	r, _ := u.gzipPool.Get().(*gzip.Reader)
+	if r == nil {
+		r, err = gzip.NewReader(bytes.NewReader(compressedRecord))
+	} else {
+		err = r.Reset(bytes.NewReader(compressedRecord))
+	}
+	if err != nil {
+		return plog.Logs{}, fmt.Errorf("failed to decompress record: %w", err)
+	}
+	defer u.gzipPool.Put(r)
+
+	type resourceKey struct {
+		owner     string
+		logGroup  string
+		logStream string
+	}
+	byResource := make(map[resourceKey]plog.LogRecordSlice)
+
+	// Multiple logs in each record separated by newline character
+	scanner := bufio.NewScanner(r)
+	for datumIndex := 0; scanner.Scan(); datumIndex++ {
+		var log cWLog
+		if err := jsoniter.ConfigFastest.Unmarshal(scanner.Bytes(), &log); err != nil {
+			u.logger.Error(
+				"Unable to unmarshal input",
 				zap.Error(err),
-				zap.Int("record_index", recordIndex),
+				zap.Int("datum_index", datumIndex),
 			)
 			continue
 		}
-		// Multiple logs in each record separated by newline character
-		for datumIndex, datum := range bytes.Split(record, []byte(recordDelimiter)) {
-			if len(datum) > 0 {
-				var log cWLog
-				err := json.Unmarshal(datum, &log)
-				if err != nil {
-					u.logger.Error(
-						"Unable to unmarshal input",
-						zap.Error(err),
-						zap.Int("datum_index", datumIndex),
-						zap.Int("record_index", recordIndex),
-					)
-					continue
-				}
-				if !u.isValid(log) {
-					u.logger.Error(
-						"Invalid log",
-						zap.Int("datum_index", datumIndex),
-						zap.Int("record_index", recordIndex),
-					)
-					continue
-				}
-				attrs := resourceAttributes{
-					owner:     log.Owner,
-					logGroup:  log.LogGroup,
-					logStream: log.LogStream,
-				}
-				lb, ok := builders[attrs]
-				if !ok {
-					lb = newResourceLogsBuilder(md, attrs)
-					builders[attrs] = lb
-				}
-				lb.AddLog(log)
-			}
+		if !isValid(log) {
+			u.logger.Error(
+				"Invalid log",
+				zap.Int("datum_index", datumIndex),
+			)
+			continue
+		}
+
+		key := resourceKey{
+			owner:     log.Owner,
+			logGroup:  log.LogGroup,
+			logStream: log.LogStream,
+		}
+		logRecords, ok := byResource[key]
+		if !ok {
+			logRecords = plog.NewLogRecordSlice()
+			byResource[key] = logRecords
 		}
-	}
 
-	if len(builders) == 0 {
-		return plog.NewLogs(), errInvalidRecords
+		for _, event := range log.LogEvents {
+			logRecord := logRecords.AppendEmpty()
+			// pcommon.Timestamp is a time specified as UNIX Epoch time in nanoseconds
+			// but timestamp in cloudwatch logs are in milliseconds.
+			logRecord.SetTimestamp(pcommon.Timestamp(event.Timestamp * int64(time.Millisecond)))
+			logRecord.Body().SetStr(event.Message)
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return plog.Logs{}, err
```

Comment:
Should this return the logs that were processed prior to the error?

Reply:
:ok_hand: I'll change this to log an error and continue, similar to decoding the records above.